File: sdks/nodejs/src/exceptions.ts
/**
 * QilbeeDB exceptions.
 */

/**
 * Base exception for all QilbeeDB errors.
 */
export class QilbeeDBError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'QilbeeDBError';
    Object.setPrototypeOf(this, QilbeeDBError.prototype);
  }
}

/**
 * Failed to connect to database.
 */
export class ConnectionError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'ConnectionError';
    Object.setPrototypeOf(this, ConnectionError.prototype);
  }
}

/**
 * Authentication failed.
 */
export class AuthenticationError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'AuthenticationError';
    Object.setPrototypeOf(this, AuthenticationError.prototype);
  }
}

/**
 * Query execution failed.
 */
export class QueryError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'QueryError';
    Object.setPrototypeOf(this, QueryError.prototype);
  }
}

/**
 * Transaction operation failed.
 */
export class TransactionError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'TransactionError';
    Object.setPrototypeOf(this, TransactionError.prototype);
  }
}

/**
 * Memory operation failed.
 */
export class MemoryError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'MemoryError';
    Object.setPrototypeOf(this, MemoryError.prototype);
  }
}

/**
 * Graph not found.
 */
export class GraphNotFoundError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'GraphNotFoundError';
    Object.setPrototypeOf(this, GraphNotFoundError.prototype);
  }
}

/**
 * Node not found.
 */
export class NodeNotFoundError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'NodeNotFoundError';
    Object.setPrototypeOf(this, NodeNotFoundError.prototype);
  }
}

/**
 * Relationship not found.
 */
export class RelationshipNotFoundError extends QilbeeDBError {
  constructor(message: string) {
    super(message);
    this.name = 'RelationshipNotFoundError';
    Object.setPrototypeOf(this, RelationshipNotFoundError.prototype);
  }
}

File: sdks/python/qilbeedb/exceptions.py
"""
QilbeeDB exceptions.
"""


class QilbeeDBError(Exception):
    """Base exception for all QilbeeDB errors."""
    pass


class ConnectionError(QilbeeDBError):
    """Failed to connect to database."""
    pass


class AuthenticationError(QilbeeDBError):
    """Authentication failed."""
    pass


class QueryError(QilbeeDBError):
    """Query execution failed."""
    pass


class TransactionError(QilbeeDBError):
    """Transaction operation failed."""
    pass


class MemoryError(QilbeeDBError):
    """Memory operation failed."""
    pass


class GraphNotFoundError(QilbeeDBError):
    """Graph not found."""
    pass


class NodeNotFoundError(QilbeeDBError):
    """Node not found."""
    pass


class RelationshipNotFoundError(QilbeeDBError):
    """Relationship not found."""
    pass


class TokenExpiredError(AuthenticationError):
    """JWT token has expired."""
    pass


class InvalidTokenError(AuthenticationError):
    """JWT token is invalid."""
    pass


class PermissionDeniedError(QilbeeDBError):
    """User lacks required permissions."""
    pass


class RateLimitError(QilbeeDBError):
    """Rate limit exceeded."""
    pass


class SecurityError(QilbeeDBError):
    """General security-related error."""
    pass
